User model

juandbc opened the pull request:
## Summary
* Add user model
* Implemented many to many relationship between user and books
* Add BookAlreadyOwnedException
* Add util ErrorConstansts class for exceptions messages
* Refactor errors messages in BookController class

## Trello Card
https://trello.com/c/2XZ1MSsw/19-crear-modelo-de-user

Commit:
Add User model
Add BookAlreadyOwnedException
Add util ErrorConstansts class
Refactor errors messages in BookController

Commit:
Add missing user id getter

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    public void removeBook(Book book) {
+        this.books.remove(book);
```
Add the check that asks if the book to be removed exists in the list

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,134 @@
+
+    @NotNull
+    @Column(nullable = false)
+    @ManyToMany
```
In the **@manytomany** tag we must indicate the type of cascade we want to use. Since by not indicating one, it will not do any function in this regard and it will bring you problems when interacting with the db.

It is usually recommended to use Casacade.ALL. But this is not always correct.
By saying **ALL**, you are saying: Persist, Merge, Refresh, Delete and Detach.

**Detach** is only so that something gets detached along the cascade when you detach an object from the persistence context manually. This almost never happens, so we won't focus on this point.

**Persist** says: If you persist the Users instance to the DB, then persist every item that this collection has in the DB as well.

**Delete** says: If you delete the Users instance from the DB, then delete every book item that this collection has in the DB as well.

**Merge** says: If you changed the in-memory values of the collection, then merge this changes to the DB so these are reflected there.

**Refresh** says: If something changed in the database values of the collection, then refresh those changes in the in-memory values.

Here is a [link](https://www.baeldung.com/jpa-cascade-types) explaining it.

In this case, we just need MERGE and REFRESH. [@ManyToMany(cascade = {CascadeType.REFRESH, CascadeType.MERGE})]
You can look at MERGE as the instruction to "merge" changes on the model back to the database.
But we're lacking the one that brings modifications iof the database to the model. And that's REFRESH :)

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,134 @@
+    @NotNull
+    @Column(nullable = false)
+    @ManyToMany
+    @JoinTable(name = "USER_BOOK",
```
The many to many relationship should work without the @JoinTable annotation since by default it defines the same thing that you have put.

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,134 @@
+
+@Entity
+@Table(name = "t_user")
+public class User {
```
When testing these enpoint with this many-to-many relationship, an error occurs when trying to display a book or a user since one contains the other and spring does not know where to stop this relationship. An infinite recursion would be generated.
To avoid this error, place this annotation on the class as follows:

```suggestion
@JsonIdentityInfo (generador = ObjectIdGenerators.PropertyGenerator.class, propiedad = "id")
public class User {
```


Place said annotation in the same way in the model book.

Commit:
Add validation to User.removeBook method
Remove redundant JoinTable annotation
Add Cascade type to User-Book relationship
Add JsonIdentityInfo annotation to User and Book models

jeymarsonwolox commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -38,13 +42,14 @@ public BookController(BookRepository bookRepository) {
     @GetMapping("/{id}")
     public Book findOne(@PathVariable Long id) {
-        return bookRepository.findById(id).orElseThrow(() -> new BookNotFoundException("Book id " + id + " not found"));
+        return bookRepository.findById(id).orElseThrow(() -> new BookNotFoundException(String.format(BOOK_NOT_FOUND,
```
excellent!!

juandbc replied:
Thanks!

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,132 @@
+    private LocalDate birthdate;
+
+    @NotNull
+    @Column(nullable = false)
```
@Column is not necessary in the relation

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,132 @@
+
+    @NotNull
+    @Column(nullable = false)
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.REFRESH})
```
I would change persist for merge.

You can look at MERGE as the instruction to "merge" changes on the model back to the database.
But we're lacking the one that brings modifications iof the database to the model. And that's REFRESH :)

juandbc replied:
My mistake, you said before that should be MERGE but I remembered like PERSIST. I'll fix it

Gianni-Nicolas commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    public void removeBook(Book book) {
+        if (books.contains(book)) this.books.remove(book);
```
We could add an exception like you did in addBook, to give it a behavior in case the book doesn't exist. What do you say?

juandbc replied:
Sound good. I'll add it

Commit:
Add exception throw in the User.removeBook method
Remove unnecessary column annotiation in User.books field
Fix cascade type in User.books relationship
